src/apps/service/apps/jobs/submissions.clj

```diff
@@ -147,12 +178,60 @@
     (async/submit-batch-jobs apps-client user ht-paths submission output-dir batch-id)
     (job-listings/list-job apps-client batch-id)))
 
+(defn- substitute-multi-input-param-values
+  [path-list-map config multi-input-param-id]
+  (assoc config multi-input-param-id (->> (get config multi-input-param-id) ;; get current multi-input path values
+                                          (map #(get path-list-map % %)) ;; expand path-list for each matched value
+                                          flatten)))
+
+(defn- pre-expand-multi-input-validation
+  [path-list-stats input-paths-by-id input-params-by-id]
+  (validate-path-list-stats "Multi-Input Path List"
+                            (config/multi-input-path-list-max-paths)
+                            path-list-stats)
+  (validate-multi-input-params (extract-path-list-params path-list-stats
+                                                         input-paths-by-id
+                                                         input-params-by-id)))
+
+(defn- config->expand-multi-input-path-lists
+  [config user input-params-by-id input-paths-by-id path-list-stats]
+  (pre-expand-multi-input-validation path-list-stats input-paths-by-id input-params-by-id)
+
+  (let [path-list-paths       (set (map :path path-list-stats))
+        path-list-map         (validate-multi-input-path-lists (util/get-path-list-contents-map user path-list-paths))
+        multi-input-param-ids (->> input-params-by-id
+                                   (filter (comp (partial = ap/param-multi-input-type) :type second))
+                                   (map first)
+                                   set)]
+    (reduce (partial substitute-multi-input-param-values path-list-map)
+            config
+            multi-input-param-ids)))
+
+(defn- process-job-config
+  [config user input-params-by-id input-paths-by-id path-stats]
+  (if-let [multi-input-path-list-stats (->> path-stats
+                                            (filter-stats-by-info-type (config/multi-input-path-list-info-type))
+                                            seq)]
+    (config->expand-multi-input-path-lists config
+                                           user
+                                           input-params-by-id
+                                           input-paths-by-id
+                                           multi-input-path-list-stats)
+    config))
+
+(defn- pre-process-submission
+  [{:keys [config] :as submission} user input-params-by-id input-paths-by-id path-stats]
+  (assoc submission :job_config (process-job-config config user input-params-by-id input-paths-by-id path-stats)))
```

Comment:
The Multi-Input Path List input parameters are expanded in a new `job_config` key of the submission.
This allows the user's original submission config to be saved to the db without all the expanded paths,
while the `job_config` key will replace the `config` in the submission that is serialized for the job services.